PEP 702: Type system support for deprecations (#2942)
python/peps: pep-0702.rst

PEP: 702
Title: Typing support for deprecations
Author: Jelle Zijlstra <jelle.zijlstra@gmail.com>
Status: Draft
Type: Standards Track
Topic: Typing
Content-Type: text/x-rst
Created: 30-Dec-2022
Python-Version: 3.12

Abstract
========

This PEP adds an ``@typing.deprecated()`` decorator that marks a class or function

> as deprecated, enabling static checkers to warn when it is used.

Reviewer: ```suggestion
as deprecated, enabling static checkers to produce a warning when it is used.
```

Author: I like the more concise current wording better

Motivation
==========

As software evolves, new functionality is added and old functionality becomes
obsolete. Library developers want to work towards removing obsolete code while
giving their users time to migrate to new APIs. Python provides a mechanism for
achieving these goals: the :exc:`DeprecationWarning` warning class, which is
used to show warnings when deprecated functionality is used. This mechanism is
widely used: as of the writing of this PEP, the CPython main branch contains
about 150 distinct deprecations. Many third-party libraries also use

> :exc:`DeprecationWarning` to mark deprecations.

Reviewer: ```suggestion
:exc:`DeprecationWarning` to mark deprecations too.
```

Author: The sentence already has "also"

* Scala `supports <https://www.scala-lang.org/api/2.12.5/scala/deprecated.html>`__
  an ``@deprecated`` annotation.
* Swift `supports <https://docs.swift.org/swift-book/ReferenceManual/Attributes.html>`__
  using the ``@available`` attribute to mark APIs as deprecated.
* TypeScript `uses <https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#deprecated>`__
  the ``@deprecated`` JSDoc tag to issue a hint marking use of
  deprecated functionality.

Several users have requested support for such a feature:

* `typing-sig thread <https://mail.python.org/archives/list/typing-sig@python.org/thread/E24WTMQUTGKPFKEXVCGGEFFMG7LDF3WT/>`__
* `Pyright feature request <https://github.com/microsoft/pyright/discussions/2300>`__
* `mypy feature request <https://github.com/python/mypy/issues/11439>`__

Specification
=============

The new ``@typing.deprecated()`` decorator can be used on a class, function or method.
With overloaded functions, the decorator may be applied to individual overloads,
indicating that the particular overload is deprecated. The decorator may also be
applied to the overload implementation function, indicating that the entire function
is deprecated.

> The decorator takes a single argument of type ``str``, which is a message that should

Reviewer: This will probably have to be a literal string, not just any string, as type checkers won't be able to obtain a string that's dynamically generated, and hence won't be able to use it in the warning.

Author: Good point, will add that

PEP: 702

> Title: Typing support for deprecations

Author: I am going with "Marking deprecations using the type system" for now.

(And happy soon-to-be-New Year!)

Author: Jelle Zijlstra <jelle.zijlstra@gmail.com>
Status: Draft
Type: Standards Track
Topic: Typing
Content-Type: text/x-rst
Created: 30-Dec-2022
Python-Version: 3.12

Abstract
========

This PEP adds an ``@typing.deprecated()`` decorator that marks a class or function
as deprecated, enabling static checkers to warn when it is used.

Motivation
==========

As software evolves, new functionality is added and old functionality becomes
obsolete. Library developers want to work towards removing obsolete code while
giving their users time to migrate to new APIs. Python provides a mechanism for
achieving these goals: the :exc:`DeprecationWarning` warning class, which is
used to show warnings when deprecated functionality is used. This mechanism is
widely used: as of the writing of this PEP, the CPython main branch contains

> about 150 distinct deprecations. Many third-party libraries also use

Author: It's code paths that lead to raising DeprecationWarning (e.g. warnings.warn, `warnings._deprecated`, whatever the equivalent in C is).

> :exc:`DeprecationWarning` to mark deprecations.

Reviewer 2: Is it easy to get some numbers for e.g. the 1000 most popular PyPI packages?

Author: https://dev.to/hugovk/how-to-search-5000-python-projects-31gk specifically, I'm currently at 472/5000 downloaded and will report back when it is finished :)

Author: Added some numbers (at least 3.5k deprecations in the top 5k packages).

  on function declarations. This powers CPython's ``Py_DEPRECATED`` macro.
* GraphQL `supports <https://spec.graphql.org/June2018/#sec-Field-Deprecation>`__
  marking fields as ``@deprecated``.
* Kotlin `supports <https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-deprecated/>`__
  a ``Deprecated`` annotation.
* Scala `supports <https://www.scala-lang.org/api/2.12.5/scala/deprecated.html>`__
  an ``@deprecated`` annotation.
* Swift `supports <https://docs.swift.org/swift-book/ReferenceManual/Attributes.html>`__
  using the ``@available`` attribute to mark APIs as deprecated.
* TypeScript `uses <https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#deprecated>`__
  the ``@deprecated`` JSDoc tag to issue a hint marking use of
  deprecated functionality.

Several users have requested support for such a feature:

* `typing-sig thread <https://mail.python.org/archives/list/typing-sig@python.org/thread/E24WTMQUTGKPFKEXVCGGEFFMG7LDF3WT/>`__
* `Pyright feature request <https://github.com/microsoft/pyright/discussions/2300>`__
* `mypy feature request <https://github.com/python/mypy/issues/11439>`__

Specification
=============

> The new ``@typing.deprecated()`` decorator can be used on a class, function or method.

Author: Pushing a slightly different wording

With overloaded functions, the decorator may be applied to individual overloads,
indicating that the particular overload is deprecated. The decorator may also be
applied to the overload implementation function, indicating that the entire function
is deprecated.

The decorator takes a single argument of type ``str``, which is a message that should
be shown by the type checker when it encounters a usage of the decorated object.

> The message must be a string literal or a string with a ``Literal`` type.

Reviewer 2: If it's a string with a `Literal` type, shouldn't it be constrained to a single value? I wouldn't want to be a type checker having to implement something like
```py
message: typing.Literal["error", "warning"]

@typing.deprecated(message)
def f(x):
    ...
```

Author: Yes, it should be a single literal, I'll rephrase. I wanted to leave the door open for something like:

```python
MESSAGE: Final = "This is deprecated for reasons x, y, z. You should use Spam instead."

@deprecated(MESSAGE)
def func1(): ...

@deprecated(MESSAGE)
def func2(): ...
```

> The message must be a string literal or a string with a ``Literal`` type.

Author: Rewriting this line to link `Final` instead.

The content of deprecation messages is up to the user, but it may include the version
in which the deprecated object is to be removed, and information about suggested
replacement APIs.

Type checkers should produce a diagnostic whenever they encounter a usage of an
object marked as deprecated. For deprecated overloads, this includes all calls
that resolve to the deprecated overload.
For deprecated classes and functions, this includes:

* References through module, class, or instance attributes (``module.deprecated_object``,

>   ``module.SomeClass.deprecated_method``, ``module.SomeClass().deprecated_method()``)

Reviewer 2: Surely also `module.SomeClass().deprecated_method` (i.e. no parens).

Author: ```suggestion
  ``module.SomeClass.deprecated_method``, ``module.SomeClass().deprecated_method``)
```

* Any usage of deprecated objects in their defining module

>   (``x = deprecated_object()`` in ``module.py``)

Reviewer 2: This one could be problematic? *Internal* uses might need to be kept around, if the deprecated object returns something whose API is not deprecated. OTOH after
```py
deprecated_alias = deprecated_object
```
using `module.deprecated_alias` elsewhere should be considered deprecated.

Author: I'm not sure why internal uses are special; e.g. when the `assertEquals` method in unittest is deprecated, unittest itself shouldn't be using it either.

* If ``import *`` is used, usage of deprecated objects from the
  module (``from module import *; x = deprecated_object()``)
* ``from`` imports (``from module import deprecated_object``)

There are some additional scenarios where deprecations could come into play:

* An object may implement a :data:`typing.Protocol`, but only if deprecated

>   methods are considered.

Author: Pushing a rewrite

* A class may use the ``@override`` decorator from :pep:`698` to assert that
  its method overrides a base class method, but the base class method is
  deprecated.

As these scenarios appear complex and relatively unlikely to come up in practice,
this PEP does not mandate that type checkers detect them.

Example
-------

As an example, consider this library stub named ``library.pyi``:

.. code-block:: python

   from typing import deprecated

   @deprecated("Use Spam instead")
   class Ham: ...

   @deprecated("It is pining for the fiords")
   def norwegian_gray(x: int) -> int: ...

   @deprecated("Only str will be allowed")
   @overload
   def foo(x: int) -> str: ...

>    @overload

Author: Yes, this is the convention in stubs. We can reconsider this in Black if the PEP is accepted but for now I'd prefer to stick with the compact syntax.

   def foo(x: str) -> str: ...

Here is how type checkers should handle usage of this library:

.. code-block:: python

   from library import Ham  # error: Use of deprecated class Ham. Use Spam instead.

   import library

   library.norwegian_gray(1)  # error: Use of deprecated function norwegian_gray. It is pining for the fiords.
   map(library.norwegian_gray, [1, 2, 3])  # error: Use of deprecated function norwegian_gray. It is pining for the fiords.

   library.foo(1)  # error: Use of deprecated overload for foo. Only str will be allowed.
   library.foo("x")  # no error

Runtime behavior
----------------

> At runtime, the decorator sets an attribute ``.__deprecated__`` on the decorated

Reviewer 3: ```suggestion
At runtime, the decorator sets an attribute ``__deprecated__`` on the decorated
```

Minor nit, but to be consistent with the usage elsewhere here and in the CPython docs

object. The value of the attribute is the message passed to the decorator.

> The decorator returns the original object. Notably, it does not issue a runtime

Reviewer 3: ```suggestion
The decorator returns the modified original object. Notably, it does not issue a runtime
```

Doesn't it return the object with the above modification? To me, it reads rather confusing to say that it modifies it and then returns the original.

Author: It returns the original object, since it is modified in place.

Reviewer 3: Right, I understand that, but it is still somewhat strange for me to read, since the object is mutated and thus not exactly the same as the original.

``DeprecationWarning``.

Type checker behavior
---------------------

This PEP does not specify exactly how type checkers should present deprecation
diagnostics to their users. However, some users (e.g., application developers
targeting only a specific version of Python) may not care about deprecations,
while others (e.g., library developers who want their library to remain
compatible with future versions of Python) would want to catch any use of
deprecated functionality in their CI pipeline. Therefore, it is recommended

> that type checker provide configuration options that cover both use cases.

Reviewer 2: Hopefully a deprecated use will not issue a warning/error when `# type: ignore` is added.

Author: Yes, the error should be `# type: ignore`able like any other type checker error. I can add an explicit mention of that if you think it's useful.

Also, unrelated typo I noticed:
```suggestion
that type checkers provide configuration options that cover both use cases.
```

Backwards compatibility
=======================

Creating a new decorator poses no backwards compatibility concerns.
As with all new typing functionality, the ``@deprecated`` decorator
will be added to the ``typing_extensions`` module, enabling its use
in older versions of Python.

How to teach this
=================

For users who encounter deprecation warnings in their IDE or type
checker output, the messages they receive should be clear and self-explanatory.
Usage of the ``@deprecated`` decorator will be an advanced feature
mostly relevant to library authors. The decorator should be mentioned

> in relevant documentation (e.g., :pep:`387` and the :exc:`DeprecationWarning`

Reviewer 3: PEP 387 (PEP-387) specifies a normative process for Python's own development, rather than being itself (user) documentation. Maybe it would be worth being explicit here what change this PEP is recommending to that one (e.g. to require `@deprecated` when possible as part of the Python deprecation process, alongside `DeprecationWarning`/`warnings._deprecated` and `.. deprecated::`/`.. deprecated-removed::` in the docs?)

Author: I think we should, but that hardly fits under "How to teach this" any more; I'll add a new section to the spec instead.

documentation) as an additional way to mark deprecated functionality.

Reference Implementation
========================

> TODO

Reviewer 2: Just suggest it to Eric Traut as a holiday project, he'll do it overnight. :-)

Author: Working on it now for pyanalyze, might also do mypy if I find time.

Rejected ideas
==============

Runtime warnings
----------------

Users might expect usage of the ``@deprecated`` decorator to issue a

> ``DeprecationWarning`` at runtime. However, this would raise a number of

Reviewer 3: ```suggestion
:exc:`DeprecationWarning` at runtime. However, this would raise a number of
```

IMO, worth being consistent here

thorny issues:

* When the decorator is applied to a class or an overload, the warning

>   would not be raised as expected.

Reviewer 3: For a class, couldn't it be raised on `__init__`, though?

Author: Or perhaps `__new__`. Correct support for all such variations feels like a huge can of worms.

* Users may want to control the ``warn`` call in more detail (e.g.,

>   changing the warning class).

Reviewer 3: That's true—in general, this would not really be useful for most CPython deprecation warnings, which use the recommended `warnings._deprecated` helper to do several useful thinks.

* ``typing.py`` generally aims to avoid affecting runtime behavior.

Users who want to use ``@deprecated`` while also issuing a runtime warning

> can use the ``if TYPE_CHECKING:`` idiom, for example:

Reviewer 2: I hope this doesn't become too popular, or we'd be better off doing this ourselves (it would be tricky to add later -- if people add other mechanisms they would get double warnings). And we _do_ have `reveal_type()`, which prints to stderr at runtime.

Author: @hauntsaninja also brought this up above. I'm now thinking of adding a `runtime=True` kwarg to the decorator that makes it also throw a runtime warning.

Reviewer 4: On the one hand, users will probably expect some runtime behaviour, and will probably be surprised if there isn't a runtime effect.

On the other hand, adding runtime behaviour feels like we might be stepping on the toes of third-party packages that have spent a lot of time thinking about this kind of thing, like https://github.com/tantale/deprecated.

(I don't really have a solution here or a strong preference, just noting that it might be an issue.)

Reviewer 3: Yeah, for what it worth, it does `raise` some potential considerations, but if people are otherwise going to be doing ugly, boilerplately and verbose things, it's better to just provide the functionality built in now, IMO.

Reviewer 5: Pyright also really doesn't like code in the `else` branch of `if TYPE_CHECKING`: https://github.com/microsoft/pyright/discussions/2330

Author: Will bring this up as a point of discussion when starting threads.

.. code-block:: python

   from typing import TYPE_CHECKING
   import functools
   import warnings

   if TYPE_CHECKING:
       from typing import deprecated
   else:
       def deprecated(msg):
           def decorator(func):
               @functools.wraps(func)
               def wrapper(*args, **kwargs):
                   warnings.warn(msg, DeprecationWarning, stacklevel=2)
                   return func(*args, **kwargs)
               wrapper.__deprecated__ = msg
               return wrapper
           return decorator

Deprecation of modules and attributes
-------------------------------------

This PEP covers deprecations of classes, functions and overloads. This
allows type checkers to detect many but not all possible deprecations.
To evaluate whether additional functionality would be worthwhile, I
`examined <https://gist.github.com/JelleZijlstra/ff459edc5ff0918e22b56740bb28eb8b>`__
all current deprecations in the CPython standard library.

I found:

* 74 deprecations of functions, methods and classes (supported by this PEP)

> * 28 deprecations of whole modules (largely due to :pep:`594`)

Reviewer 2: Hm. This does make me think that maybe we should try a bit harder to come up with something for modules. (Maybe at least when they are submodules of a package? Surely that must be the most common case for 3rd party packages.)

Author: I feel like the suggested implementation for whole-module deprecations a few paragraphs down is sound, so we can add it if we think it's sufficiently useful. I'll bring this up when I open discussion threads to get more feedback on whether it is useful.

This number might be a bit of a distortion because CPython happens to be in the middle of lots of PEP 594 deprecations, so the number of whole-module deprecations is higher than it would usually be.

Reviewer 2: > This number might be a bit of a distortion because CPython happens to be in the middle of lots of PEP 594 deprecations, so the number of whole-module deprecations is higher than it would usually be.

I actually expect that more ancient, poorly-maintained modules in the stdlib will eventually get the axe.

Author: Will bring this up (too) as a point of discussion when starting threads.

* 9 deprecations of function parameters (supported by this PEP through
  decorating overloads)
* 1 deprecation of a constant

> * 38 deprecations that are not easily detected in the type system (for

Reviewer 2: ```suggestion
* 38 deprecations that are not easily detectable in the type system (for
```

  example, for calling :func:`asyncio.get_event_loop` without an active
  event loop)

Modules could be marked as deprecated by adding a ``__deprecated__``
module-level constant. However, the need for this is limited, and it
is relatively easy to detect usage of deprecated modules simply by
grepping. Therefore, this PEP omits support for whole-module deprecations.
As a workaround, users could mark all module-level classes and functions
with ``@deprecated``.

For deprecating module-level constants, object attributes, and function
parameters, a ``Deprecated[type, message]`` type modifier, similar to
``Annotated`` could be added. However, this would create a new place
in the type system where strings are just strings, not forward references,
complicating the implementation of type checkers. In addition, my data
show that this feature is not commonly needed.

Acknowledgments
===============

A discussion in the typing-sig meetup group led to useful feedback on this

> proposal.

Reviewer 3: Could you link this?

Author: No, it was a Zoom call.

Reviewer 3: Okay, then perhaps that could be clarified? I read this as a discussion on the Typing-SIG mailing list. E.g.

```suggestion
A Zoom call in the typing-sig meetup group led to useful feedback on this
proposal.
```

or

```suggestion
A Zoom discussion in the typing-sig meetup group led to useful feedback on this
proposal.
```

Author: Pushed a change

Specification
=============

A new decorator ``@deprecated()`` is added to the :mod:`typing` module. This
decorator can be used on a class, function or method to mark it as deprecated.
With overloaded functions, the decorator may be applied to individual overloads,
indicating that the particular overload is deprecated. The decorator may also be
applied to the overload implementation function, indicating that the entire function
is deprecated.

The decorator takes a single argument of type ``str``, which is a message that should
be shown by the type checker when it encounters a usage of the decorated object.
The message must be a string literal or be statically resolvable to a single string

> (e.g., a :data:`typing.Final` variable of type ``str``).

Reviewer 2: Can't say I believe this is the right constraint either. It's totally possible to have something of type `Final[str]` whose value cannot be determined by the type checker -- that type just means that it cannot be reassigned, but the initial value could still be dynamically computed.

Reviewer 3: What about

```suggestion
(e.g., a :data:`~typing.Final` variable of type ``str`` assigned a string literal).
```

Reviewer 2: Presumably for pyright that wouldn't be a problem, since it already tracks literal values. In mypy this will probably require something new, since it doesn't do this. No idea about Pyre, pytype and others. Maybe we should leave this discussion open for the more public debate of the PEP's merits and details. A base position to retreat to might be that it has to be a string literal? PEP 675 (`LiteralString`) comes to mind.

Author: I'll just go back to only string literals; type checkers can relax the requirements if they wish.

deprecated functionality in their CI pipeline. Therefore, it is recommended
that type checkers provide configuration options that cover both use cases.

Deprecation policy
------------------

> CPython's deprecation policy (:pep:`387`) will require that new deprecations

Reviewer 2: This is what's being _proposed_, right?

Author: Yes, should that be made more obvious somehow?

Reviewer 2: I had to read it twice before I was sure of it. Maybe:
```suggestion
We propose that CPython's deprecation policy (:pep:`387`) is updated to require that new deprecations
```
